### PR TITLE
router: Change parser for user defined headers to allow for single string arguments

### PR DESCRIPTION
**Pull request by curiouserrandy**
*Description*:
Change the parser for user defined header configuration to allow for single (not array) string arguments for variables, e.g. '%PER_REQUEST_STATE("reverse.dns.data.name")%').

*Risk Level*: Low

*Testing*: This code is currently tested at the level above its implementation.  I'm relying on existing tests for lack of regression of functionality, and am adding new tests for the consumers of this new functionality in the PR which creates those consumers. 

*Docs Changes*: None in this PR.

*Release Notes*: None from this PR.

### CODE REVIEW

**Comment by curiouserrandy** [comment]
For context as to how this change is intended to be used, see https://github.com/envoyproxy/envoy/compare/master...curiouserrandy:DynamicUDHOnly

**Comment by curiouserrandy** [comment]
@zuercher : Would you be willing to take a look at this?

**Comment by zuercher** [comment]
The parser already supported single string arguments, but as I looked at the original to better understand your change, I realize a bug was introduced at some point. Could you add a test case in header_formatter_test's TestFormatWithUpstreamMetadataVariable like this:

```
  // Top-level value with paren.
  testFormatting(request_info, "UPSTREAM_METADATA([\"namespace\", \"(key)\"])", "value");
```
 (It can go near the existing "Top-level value" case.)

That should fail with the existing parser because the close paren causes an incomplete set of args to be passed to the RequestInfoHeaderFormatter. I think your change actually fixes that bug.

Beyond fixing that bug, this seems to provide the ability to quote the single argument
```
%VAR("xyz")%
```
versus the current style:
 ```
%VAR(xyz)%
```

Looking at the change you referenced, it seems like you're expecting reverse-DNS style names. Is the additional complexity necessary?

**Comment by curiouserrandy** [comment]
a) I will add the test case.  Agreed about the bug introduction at some point and the fix. I think my change also has more protection against starting/ending mismatches (starting parsing an array of arguments and ending as if it was singular) but I didn't try to come up with an example test case. 

b) Yes, this is to allow quoting the single argument with strings.  If you want, I'll switch over to using just the current style, but given that the string is going to be used as a lookup in a data structure, that didn't "feel right" to me.  But consistency may override that feeling and if you want me to change the expected format I will.  What do you think?

c) The specification of reverse DNS names for the consumer use case was requested during the design by @htuch .  Having said that, there shouldn't be anything in this PR that mentions that (there's no reason for the parser to know what the string format is) and there should be nothing in the consuming PR (the branch I pointed to in my first comment) that enforces it.  Are either of those two statements incorrect?

**Comment by curiouserrandy** [comment]
Followup: What exactly are you wanting to test with the test line you suggested?  It fails in exactly the same way with and without my change (expected_output ("value") != formatted_string ("")), and as I mentioned, my perception of the bug that I fixed was around starting and ending strings and list inside the parens differently.

**Comment by zuercher** [comment]
Huh. Leave it out the test case. I'll see about adding myself.

I would like to see the single quoted string parsing exercised, particular of malformed arguments. I think you'd either have to refactor parseInternal to allow it to be tested separately from RequestInfoHeaderFormatter and PlainHeaderFormatter or else write tests using START_TIME.

For b/c above, I'm ok with the quoted single string argument. I think it allows exact control of leading/trailing spaces, which someone will want eventually. I agree that this PR needn't deal with the exact format of the string, I was just exploring the reasons for adding the quoted string argument. And it needn't be enforced in your forthcoming PR either. It's just convention.

**Comment by curiouserrandy** [comment]
(I did look into testing parseInternal separately, and it's pretty easy to make the function visible to tests, but it's not very easy to probe the HeaderFormatter it returns, so you're basically back to asking the formatter to format something and looking at the results, which is what's there currently.)

File: test/common/router/header_formatter_test.cc [hunk 1]
```diff
@@ -269,6 +269,7 @@ TEST(HeaderParserTest, TestParseInternal) {
       {"%UPSTREAM_METADATA([\"ns\", \n \"key\"])%", {"value"}, {}},
       {"%UPSTREAM_METADATA( \t [ \t \"ns\" \t , \t \"key\" \t ] \t )%", {"value"}, {}},
       {"%START_TIME%", {"2018-04-03T23:06:09.123Z"}, {}},
+      {"%START_TIME(\"testing\")%", {"\"testing\""}, {}},
```
Comment (zuercher) [inline test/common/router/header_formatter_test.cc]: Should parseInternal strip the quotes? I would expect that either `%START_TIME(xyz)%` or `%START_TIME("xyz")%` would produce the same output.

File: test/common/router/header_formatter_test.cc [hunk 2]
```diff
@@ -298,6 +299,12 @@ TEST(HeaderParserTest, TestParseInternal) {
        {"Invalid header configuration. Un-terminated variable expression 'VAR after'"}},
       {"% ", {}, {"Invalid header configuration. Un-terminated variable expression ' '"}},
 
+      // Mismatched array beginning and termination.
+      {"%START_TIME([\"testing)%",
```
Comment (zuercher) [inline test/common/router/header_formatter_test.cc]: I would do this with UPSTREAM_METADATA, but I think it's fine to add the test case.

File: test/common/router/header_formatter_test.cc (hunk 1 above)
Comment (curiouserrandy) [reply test/common/router/header_formatter_test.cc]: I can see why you'd say that, but from reading the code it looks to me like a fairly big change to the responsibilities of parseInternal() and hence a reasonably large refactor to shift responsibilities around.  I'd describe what parseInternal currently does as "Find the beginning and end % characters for a range of formats, and give RequestInfoHeaderFormatter the string between those % characters".  This change would imply that parseInternal would either edit that string in a specific way or would give RequestInfoHeaderFormatter a list of tokens based on its parsing.  Is that really what you want?

To put it another way, if we want %START_TIME(xyz)% and %START_TIME("xyz")% to produce the same output, given the current structure of the code the obvious place to do that is in AccessLog::AccessLogFormatParser.  Would you like me to make that change there?

File: test/common/router/header_formatter_test.cc (hunk 2 above)
Comment (curiouserrandy) [reply test/common/router/header_formatter_test.cc]: Done

**Comment by curiouserrandy** [comment]
I didn't see a response to you about my question about unmatched terminal quotes; I just want to confirm that means that you're ok with the current behavior?  (This is specifically around the test case I didn't add because it doesn't fail:

{"%START_TIME(testing\"])%", {"testing\"]"}, {}},

)

**Comment by zuercher** [comment]
I'm fine with leaving out that test case.

**Comment by curiouserrandy** [comment]
Just confirming that you're not waiting on me?  I believe I've responded to all your comments (although only executed on one of them).

File: test/common/router/header_formatter_test.cc (hunk 1 above)
Comment (zuercher) [reply test/common/router/header_formatter_test.cc]: Sorry, my browser has been acting weirdly. Thought I responded to this, but it's not here, so I guess not.

Handling the quotes in AccessLogFormatParser will affect %START_TIME% when used in access and upstream logs. I wanted to ask @dio what he thought about all this.

Comment (curiouserrandy) [reply test/common/router/header_formatter_test.cc]: @dio: Any thoughts here?

Comment (dio) [reply test/common/router/header_formatter_test.cc]: Followed the thread above, I think I agree that `%START_TIME(xyz)%` and` %START_TIME("xyz")%` should produce the same output, since this additional enclosing `"` to an arg, as mentioned above, is to state that that arg is a string.

Comment (curiouserrandy) [reply test/common/router/header_formatter_test.cc]: Happy to do it, and I'm inclined to do it as part of this PR; if I'm allowing new formats I should probably handle them as they are going to be handled in the future.

Just to give a heads up in case anyone wants to weigh in before I upload the next revision: I think this means I need to modify parsing for %START_TIME% (access log and custom headers) and %DYNAMIC_METADATA% (custom headers only) s.t. if a single string argument is surrounded by quotes it will strip the quotes and use what's inside the quotes as the format string.  

I debated whether I needed to modify %REQ/RESP/TRAILER% parsing and decided that that would be more complex and arguably wouldn't make sense.  Though I could see an argument that quotes should be stripped from the X & Y arguments separately, I couldn't see one for stripping them from the combination.

Comment (curiouserrandy) [reply test/common/router/header_formatter_test.cc]: Actually, I'm going to skip DYNAMIC_METADATA, as it's not really a string argument; like REQ/RESP it's several string arguments with delimiting characters.

**Comment by curiouserrandy** [comment]
I believe this is ready for another round of review.

**Comment by zuercher** [comment]
I think the trouble with doing it this way is that START_TIME (or any other single-arg formatter) now needs to unescape backslashes.

I think the right way to do this is to have parseRequestInternal accept three types of arguments and completely decode them, passing the results as an array/vector of args to RequestInfoHeaderFormatter.
1. %XXX(bare_string)% invokes RequestInfoHeaderFormatter with {"bare_string"} (without handling escaped characters -- this is basically backwards compatibility for START_TIME).
2. %XXX("quoted_string")% invokes RIHW with {"quoted_string"} (handling escapes)
3. %XXX(["json", "array"])% invokes RIHW with {"json, "array"}.

This would involve moving the JSON parser stuff into parseRequestInternal and removing the backslash escapes from quoted strings (which one might be able to use the JSON parser for as well). Additionally, would have to rework the RequestInfoHeaderFormatter to take arguments.

I recognize doing all that is a much bigger change than you originally set out to make. It seems that for the feature you're trying to add, just leaving the existing support for 1 would work, correct? If so, I'm perfectly fine if you want to set this aside (or abandon it completely) and I'll file a ticket to do the bigger refactor.
